### PR TITLE
GAWB-2855: Onix Billing Report

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses
https://broadinstitute.atlassian.net/browse/GAWB-2855

## Notes
This PR adds the ability for Orchestration to write/update a spreadsheet file to a user's google drive. The data is canned Free Trial data.

Currently, this assumes the value of the Onix user agreement value, and will need to be updated once #566 lands to take advantage of the new user trial status value. 

For a future story, I'd love for the update endpoint to be smarter. Currently, it replaces the data that it writes to. It would be much nicer if it first checked the existing values, update those in their current rows, and then append any other data that doesn't already exist (see https://broadinstitute.atlassian.net/browse/GAWB-2963).

This PR requires that we enable two APIs (Google Drive and Google Sheets APIs) in our google projects for each environment. I have done this for dev, but it needs to be done for all other environments and we need to make sure this gets onto Devops' radar so their FC project setup documentation can be updated.

---
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've followed [the instructions](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [x] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [x] I've updated the [FISMA documentation](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [x] Get two thumbsworth of review and PO signoff if necessary
- [x] Verify all tests go green
- [x] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/swagger/api-docs.yaml
Comment (rushtong) [inline src/main/resources/swagger/api-docs.yaml:1726]: I am considering a better response for these endpoints. I think it makes sense to provide a link to the created/updated spreadsheet.

Comment (davidangb) [reply src/main/resources/swagger/api-docs.yaml:1726]: if you can return the link, 200 OK is fine; if there's no body to the response, maybe 201 Created or 202 Accepted or maybe even 204 No Content

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: is this call to `getUserProfile` necessary? If the user made it this far, past our auth proxy, can't you assume they have a google account?

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: does this overwrite existing data in the spreadsheet? In other words, if some campaign manager manually adds comments to a cell or formatting or something ... will those be lost?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: instead of two queries, why not just update `projectReport` to return all projects?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: I realize this isn't in the requirements, but now that I'm looking at the code: it probably makes sense to add termination date here too.

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala
Comment (davidangb) [inline src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala:75]: it looks like this supports the ability to have multiple spreadsheets. Do we really want that? I was picturing a single spreadsheet, and all involved know its location.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: We need to build a new credential object to make the spreadsheet API call. To do that, we need the bearer token. I didn't see any other way to get that other than going through the google API for that.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: It does overwrite the data, but does not change cell formatting. The way I envision this working would be that orch writes to the same worksheet columns every time, and all charts, additional columns (outside of what we update), etc. would remain in place. I think a best practice would be that the worksheet orch touches is never modified by users and instead, they make their changes in other worksheets.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: I could do that - did not want to muck with existing queries that used `projectReport`

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: Yeah, easy enough to do. I was thinking we would certainly want more info.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: Additionally, it might make sense to write a set of instructions in one worksheet, data in another, and let the user do what they want in all of the others.

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiService.scala:75]: The issue is that we need to write an initial spreadsheet to *someone's* google drive since FC doesn't have one. That person would then have to be the one to manage sharing. Orch would never change that, unless we want to add that kind of functionality.

Side note - I did think that perhaps the initial campaign manager could supply a list of email addresses that we add as shared users - not sure if that's technically possible yet.

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala
Comment (davidangb) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: But you're building the credential from the `userInfo` that is passed in:
```
val credential = new GoogleCredential().setAccessToken(userInfo.accessToken.token)
```

The `userResponse` is only used to make the `responseStr`, and `responseStr` is only used for logging.

Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala]: That will simplify things, yes, thank you!

**Comment by kyuksel** [comment]
The corresponding JIRA tickets also includes `whether or not the user accepted Onix's user agreement` in the list of things to report. The endpoint to persist that information is not yet in place I believe. Is there a stub for that call in this PR?

**Comment by kyuksel** [comment]
I couldn't hit the endpoints on Swagger due to authorization errors. When I run Orch locally with this branch and hit `Authorize` button on top right of the Swagger page, I don't see the second `Authorize` button on the upcoming dialogue box (as shown below) anymore. Am I missing something?

authorizebuttonmissing

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (kyuksel) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: `TrailServiceSupport` trait seems to exist for Onix spreadsheet reporting purposes only. Does it make sense to rename it to reflect that?

File: src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/ElasticSearchTrialDAOSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/ElasticSearchTrialDAOSpec.scala]: I couldn't understand why the lines that are commented out were left over.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: Instead of the `campaignManager` and `nonCampaignManager` `String` literals, consider using `manager` and `unauthorizedUser` defined in `TrialApiServiceSpec` object so we have one place to modify them as necessary.

File: src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/ElasticSearchTrialDAOSpec.scala
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/ElasticSearchTrialDAOSpec.scala]: I'll fix that ... it's carry over from before I changed what `projectReport` does. That change was one of David's suggestions.

**Comment by rushtong** [comment]
@kyuksel - With respect to the authentication, it's a swagger UI weirdness. You have to scroll within the list of auth scopes to get to the auth button. Definitely a swagger thing - and not intuitive.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: 👍  Will do.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: The intention is that this supports all of `TrialService`. See `LibraryServiceSupport` and `PermissionsSupport` for how we approach this idea.

File: src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/ElasticSearchTrialDAOSpec.scala
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/integrationtest/ElasticSearchTrialDAOSpec.scala]: ... re-read your comment. They were left there intentionally by David so other developers would be able to see all of the potential trial projects that could be returned. I will update the comment to be more clear.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala
Comment (kyuksel) [inline src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: Note that if you use the `manager` val defined in the companion object, you should be able to remove the case above since `"manager"` is defined as one of the `"trial_managers"` above. In my previous PR, I meant to turn 

```scala 
"trial_managers" -> Seq("manager")
```
into 
```scala
"trial_managers" -> Seq(manager)
```
but forgot. It will be in my next PR and you're welcome to do that here, too.

**Comment by kyuksel** [comment]
> With respect to the authentication, it's a swagger UI weirdness. You have to scroll within the list of auth scopes to get to the auth button. Definitely a swagger thing - and not intuitive.

Doh!

Getting past that, my authorization attempt failed with

screen shot 2017-12-08 at 1 06 57 pm

**Comment by rushtong** [comment]
@kyuksel - Technically, it's not a failure, just an extra step to go through. Click the advanced -> proceed link and it will work. I will work with devops on seeing if we can get Google to verify this when running locally.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: Ugh, this is tougher. Doing that breaks other tests. Looking more closely into that.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (kyuksel) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: We also have `Trial.scala` that I thought served the same purpose albeit it's an `object`. Not a big deal, though.

Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: The `Trial` class looks like a holder for all of the trial-related case classes/objects - the spreadsheet utilities didn't seem to fit there. If this does become a something more broadly used, it should probably go into a google support class instead of a trial support class.

**Comment by kyuksel** [comment]
Hitting `/api/trial/manager/report` returns something like 

```json
{
  "spreadsheetUrl": "https://docs.google.com/spreadsheets/d/1JOdoq34pQvlnbJa-LpVnvQ"
}
```
and then the parameter to `/api/trial/manager/report/{spreadsheetId}` seems to be the last segment on the above URL, which wasn't obvious to me. I wonder if we can clarify it on the Swagger comments or something.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala
Comment (kyuksel) [reply src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: Probably not worth it if it ends up taking too much effort.

Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: I did get it working ... it was my own stupidity.

File: src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala
Comment (rushtong) [reply src/test/scala/org/broadinstitute/dsde/firecloud/webservice/TrialApiServiceSpec.scala]: 👍 Got both in, thank you!

**Review by kyuksel — APPROVED** [review]

**Comment by rushtong** [comment]
Moved PR from `develop` to `free_tier` branch

**Comment by kyuksel** [comment]
Re-verified that it still works 👍  

A few suggestions that shouldn't necessarily hold up the merge:

1. Return `404` instead of `500` when a user enters an ID that doesn't correspond to an existing document.

2. In the aforementioned scenario, extract the `Resource Not Found` message from the JSON and return that instead of the entire JSON.

3. On `/api/trial/manager/report`'s Swagger, note what part of the response URL will be the ID for subsequent document update requests.

**Comment by aednichols** [comment]
@rushtong - https://github.com/broadinstitute/firecloud-orchestration/pull/566 is now on dev.

**Comment by rushtong** [comment]
@aednichols - Re-based PR to`develop` and needed to do a minor merge fixup.

File: src/main/resources/swagger/api-docs.yaml
Comment (aednichols) [inline src/main/resources/swagger/api-docs.yaml:1752]: I am a bit worried about users clobbering existing spreadsheets of theirs... maybe we should say something like
> This endpoint replaces the contents of the specified Google spreadsheet with the free trial report. The spreadsheetId can be found in the URL of the link. For example, in the link "https://docs.google.com/spreadsheets/d/long-random-string/edit", the string "long-random-string" is the "spreadsheetId" value. (You may wish to pre-create a blank spreadsheet to use this feature.)

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (aednichols) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: This works as expected, but I've heard that we now prefer imports together at the top for style/consistency

**Review by aednichols — COMMENTED** [review]
Looking good, I have to run to an interview recap so this is my feedback so far:

- This is really cool to use! I tested with the Swagger on one display and the spreadsheet in another, and the real-time update is great.
- Whenever we display `Thu, 1 Jan 1970 00:00:00` in the report, compsci people know that means `0` or `null`, but maybe we should say `(No data)` for the business people?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala
Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialServiceSupport.scala]: I kept it local so as not to interfere with other scala-y things but will test out moving it.

File: src/main/resources/swagger/api-docs.yaml
Comment (rushtong) [reply src/main/resources/swagger/api-docs.yaml:1752]: Yes, I have that task as a separate story (see https://broadinstitute.atlassian.net/browse/GAWB-2963). The current approach has Moran's blessing, as flawed as it is.

**Comment by rushtong** [comment]
+1 to showing no data for `Instant`-zero cases.

**Review by aednichols — APPROVED** [review]

**Pull request merged**
